Fast manual file sync for the back end

m-vo opened the pull request:
This PR makes the manually file sync in the back end (basically `DC_Folder::sync()`) use the new DBAFS service.

With this, there is no difference anymore between using the back end or the `contao:filesync` command on the CLI. Technically, we could deprecate the old `\Contao\Dbafs::syncFiles()` now, but we might want to leave this until the back end is fully migrated and we consider the DBAFS manager stable.

I simplified the report a bit to fit the new `ChangeSet` (for instance removed "unchanged") but reused styles and labels where possible:

![image](https://user-images.githubusercontent.com/5305677/150654594-7ce5496d-40de-46f0-ad93-11a09c741a42.png)

As the label that's summing up  the results previously contained more placeholders, I adjusted our translator to gracefully handle the transition in https://github.com/contao/contao/commit/18b5535fbfcf8a6385b6f1d5d4dda38181986512. The solution is not perfect but probably enough. If we want that, I could also extract this part into another PR.

m-vo committed:
use the contao.filesystem.dbafs_manager when manually synchronizing in the back end

m-vo committed:
adjust the translator to gracefully handle additional parameters

ausi reviewed (commented):
✅ Approved apart from the Translator changes. I think we should discuss them in a different PR and you can add the two extra parameters fixed to `0`. Or create a new label for the new thing in this PR.

ausi committed:
Revert 18b5535fbfcf8a6385b6f1d5d4dda38181986512

ausi committed:
Fix translation backwards compatibility

ausi committed:
Fix cut-off text

ausi commented:
I reverted the Translator changes in 689f34869a01ded4773bb424943339d6141fe6b0, fixed BC with the old label in 8dd3ec1e2fa205dc18ac9951f5cfcb9f911cad5e and fixed a style bug in 80d7e37dc5884ba9c36546256f868d1762cd7dde

The old labels now look like this (until the new translation is added)
![Bildschirmfoto 2022-01-28 um 21 48 00](https://user-images.githubusercontent.com/367169/151620234-30150dee-8dc9-4785-ae81-2ed44434a227.png)

New label:
![Bildschirmfoto 2022-01-28 um 21 48 56](https://user-images.githubusercontent.com/367169/151620186-670cd826-9805-43c9-a8e8-2d288e2f519f.png)

m-vo commented:
👍 to the commits by @ausi. Thanks for adjusting!

leofeyer commented on core-bundle/src/Resources/contao/languages/en/tl_files.xlf:
```diff
@@ -129,7 +129,7 @@
       </trans-unit>
       <trans-unit id="tl_files.syncResult">
-        <source>&lt;strong&gt;Result:&lt;/strong&gt; %s added | %s modified | %s unchanged | %s moved | %s deleted</source>
+        <source>&lt;strong&gt;Result:&lt;/strong&gt; %s added | %s updated | %5$s deleted</source>
```
`%5$s`?

ausi replied:
That means use the 5th parameter instead of the 3rd. This is to make the label compatible with older translations. (because the 5th parameter was used for “deleted” files previously).

https://www.php.net/manual/en/function.printf.php

> An integer followed by a dollar sign $, to specify which number argument to treat in the conversion.

leofeyer replied:
So does this mean that we are no longer tracking modified, unchanged and moved files?

ausi replied:
Modified and moved is consolidated into updated AFAIK. The count of unchanged files is probably not that valuable as it is just the count of all files minus updated and added.

leofeyer committed:
Merge branch '4.13' into filesystem/be-filesync

leofeyer committed:
CS

leofeyer reviewed (changes requested):
There are two issues with this PR:

1. `.DS_Store` files are now added. They haven‘t been before.

<img width="875" alt="" src="https://user-images.githubusercontent.com/1192057/152515693-77f7c02e-c6bf-4b49-871c-1a8a54f57655.png">

2. The "Show unchanged" option is no longer there.

<img width="875" alt="" src="https://user-images.githubusercontent.com/1192057/152515812-bce993a5-776f-42d2-ba1b-02783333037f.png">

Toflar commented:
I never understood "Show unchanged". It's just the rest and I don't care about them because they are unchanged :D

leofeyer commented:
This is not about personal preferences. We cannot remove a feature just because you never understood it. 🤷‍♂️

Toflar commented:
This is not about personal preference? The information itself is objectively pointless.

m-vo commented:
> .DS_Store files are now added. They haven‘t been before.

Then that was either wrong before or should be adjusted in the DBAFS service in general. But that's unrelated to this PR in my perspective.

> The "Show unchanged" option is no longer there.

The unchanged files (basically all files minus changed) aren't listed anymore, so this option was removed. Listing unchanged files is not helpful in my eyes and the option a good indicator that you probably wanted to hide them anyway. 🙃 The new change set also does not contain this information.

> Modified and moved is consolidated into updated AFAIK.

Yes. We can differentiate between path changed/hash changed/both changed, but as these could overlap (if both change) the individual counts would be misleading.

leofeyer reviewed (changes requested):
As I said, we cannot remove existing features just because they are "not helpful in your eyes" or you "don't care about it". I cannot merge the PR this way.

ausi commented:
> > .DS_Store files are now added. They haven‘t been before.
> 
> Then that was either wrong before or should be adjusted in the DBAFS service in general.

Needs to be adjusted in DBAFS IMO. A configurable pattern of ignored files maybe? Currently we skip all files that begin with a dot: https://github.com/contao/contao/blob/f577f1ebadebefd88eb44177ab4115a80fcad55a/core-bundle/src/Resources/contao/library/Contao/Filter/SyncExclude.php#L34

> As I said, we cannot remove existing features just because they are "not helpful in your eyes"

We do this all the time, e.g. removing Google Fonts from the layout.

So I think we should discuss the usefulness of the counts and implement what we agree on.

My personal favorite would be `N added | N modified | N moved | N deleted` as it can be helpful, if for example I do some renames on the server, that I can see that nothing was “modified” but only „moved“.

m-vo commented:
> My personal favorite would be N added | N modified | N moved | N deleted as it can be helpful, if for example I do some renames on the server, that I can see that nothing was “modified” but only „moved“.

Would you then *not* count a file as modified that was also moved or vice versa? Otherwise the individual sums would not add up.

ausi commented:
> Would you then _not_ count a file as modified that was also moved or vice versa?

If a file was moved AND modified I’d count them in both categories. I would not expect the sums to have to add up based on the shown counts.

m-vo committed:
distinguish between modified and moved files in the result counts

ausi commented:
Voting might help us out here ☺️

@contao/developers please vote:
- 🎉 for `N added | N modified | N moved | N deleted` the current state of this PR.
- 🚀 for `N added | N modified | N unchanged | N moved | N deleted` as it is now in 4.12 including unchanged files.

ausi commented:
My voting questions were not optimal 🙈

@contao/developers the real question is:

- 🎉 for removing the “Show unchanged” option (as currently in this PR)
- 🚀 for keeping/reimplementing the “Show unchanged” option with the list of all files

leofeyer commented:
I‘m still not a fan of removing the option but at least I can blame you now if it upsets the users. 😛

leofeyer committed:
Merge branch '4.13' into filesystem/be-filesync

leofeyer commented on core-bundle/src/Resources/views/Backend/be_filesync_report.html.twig:
```diff
@@ -0,0 +1,44 @@
+    {% endfor %}
+
+    {# Modified resources #}
+    {% for path, updated in modified %}
```
It seems you are not using the `updated` variable?

m-vo replied:
Would you prefer using `modified|keys`?

leofeyer replied:
Not sure. Isn‘t it `{% for path in modified %}` then?

leofeyer replied:
Ah, I see. Associative array. Never mind then. 👍

leofeyer commented on core-bundle/src/Resources/views/Backend/be_filesync_report.html.twig:
```diff
@@ -0,0 +1,44 @@
+    {% endfor %}
+
+    {# Deleted resources #}
+    {% for deleted, type in change_set.itemsToDelete %}
```
It seems you are not using the `type` variable?

leofeyer commented:
This still needs to be fixed:

> 1. `.DS_Store` files are now added. They haven‘t been before.
> 
> <img alt="" width="875" src="https://user-images.githubusercontent.com/1192057/152515693-77f7c02e-c6bf-4b49-871c-1a8a54f57655.png">

These are the options we are using now:

https://github.com/contao/contao/blob/be0d1bbff682506f0f9ed9d71425fc91c45c5cea/core-bundle/src/Resources/contao/library/Contao/Dbafs.php#L129

We should not change the behavior IMHO.

m-vo commented:
> .DS_Store files are now added. They haven‘t been before.

see #4111

leofeyer commented:
Thank you @m-vo.
